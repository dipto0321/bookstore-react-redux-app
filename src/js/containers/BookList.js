import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { switchAddBookMode, removeBook, switchProgressUpdate } from '../actions/books';
import { changeFilter } from '../actions/filter';
import Book from '../components/Book';
import CategoryFilter from '../components/CategoryFilter';
import ProgressUpdateModal from '../components/ProgressUpdateModal';

export const BookList = (
  {
    books,
    changeFilter,
    filter,
    removeBook,
    switchAddBookMode,
    progressUpdateMode,
    switchProgressUpdate,
  },
) => (
  <div>
    <ProgressUpdateModal
      progressUpdateMode={progressUpdateMode}
      switchProgressUpdate={switchProgressUpdate}
    />
    <div className="booklist-controls">
      <CategoryFilter
        filter={filter}
        handleChange={optionValue => changeFilter(optionValue)}
      />
      <button
        className="add-book-btn btn-sm"
        type="button"
        onClick={switchAddBookMode}
      >
          Add Book
      </button>
    </div>
    <div>
      {books.map(book => (
        <Book
          key={book.id}
          book={book}
          handleRemove={id => removeBook(id)}
        />
      ))}
    </div>
  </div>
);

const mapStateToProps = state => ({
  books: state.books.filter(book => (state.filter === '' ? true : book.category === state.filter)),
  filter: state.filter,
  progressUpdateMode: state.progressUpdateMode,
});

BookList.propTypes = {
  books: PropTypes.instanceOf(Object).isRequired,
  changeFilter: PropTypes.func.isRequired,
  filter: PropTypes.string.isRequired,
  removeBook: PropTypes.func.isRequired,
  switchAddBookMode: PropTypes.func.isRequired,
  progressUpdateMode: PropTypes.instanceOf(Object).isRequired,
  switchProgressUpdate: PropTypes.func.isRequired,
};

export default connect(
  mapStateToProps,
  {
    removeBook, changeFilter, switchAddBookMode, switchProgressUpdate,
  },
)(BookList);
