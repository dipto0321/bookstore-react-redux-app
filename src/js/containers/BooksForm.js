import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import categories from '../data/bookCategories';

const BooksForm = () => {
  return (
    <form>
      <input type="text"/>
      <select>
        {
          categories.map(category => (
            <option
              key={category}
              value={category}
            >
              { category }
            </option>
          ))
        }
      </select>
      <button 
        type="submit"      
      >
        Add Book
      </button>
    </form>
  )
};

export default BooksForm;
