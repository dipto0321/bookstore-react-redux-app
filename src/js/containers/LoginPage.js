import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { login } from '../thunks/currentUser';

export const LoginPage = ({ login, error, history }) => {
  const [emailOrUsername, setEmailOrUsername] = useState('');
  const [password, setPassword] = useState('');
  const [errorMessage, setErrorMessage] = useState(error);

  useEffect(() => {
    setErrorMessage(error);
  }, [error]);

  const loginUser = (e) => {
    e.preventDefault();
    login({ email_or_username: emailOrUsername, password })
      .then(() => {
        setEmailOrUsername('');
        setPassword('');
        setErrorMessage(error);
        history.push('/');
      }).catch(() => {});
  };

  return (
    <form>
      {
        error && <div className="error">{errorMessage}</div>
      }
      <div>
        <label htmlFor="email">Username/Email:</label>
        <input
          type="email"
          id="email"
          value={emailOrUsername}
          onChange={e => setEmailOrUsername(e.target.value)}
        />
      </div>
      <div>
        <label htmlFor="password">Password:</label>
        <input
          type="password"
          id="password"
          value={password}
          onChange={e => setPassword(e.target.value)}
        />
      </div>
      <button
        type="submit"
        onClick={loginUser}
      >
        Login
      </button>
    </form>
  );
};

const mapStateToProps = state => ({
  error: state.error,
});

export default connect(mapStateToProps, { login })(LoginPage);
