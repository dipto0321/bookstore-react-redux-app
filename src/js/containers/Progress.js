import React from 'react';
import CircularProgressbar from 'react-circular-progressbar';

const Progress = () => {
  const percentage = 66;
  return (
    <div className="progress-body">
      <div className="progress-bar">
        <CircularProgressbar percentage={percentage} />
      </div>
      <div className="progress-percentage">
        <p className="percent-complete">{`${percentage}%`}</p>
        <p className="completed">Completed</p>
      </div>
    </div>
  );
};

export default Progress;
