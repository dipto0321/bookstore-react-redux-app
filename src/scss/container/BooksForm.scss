.add-new-title {
  font-family: "Montserrat", sans-serif;
  font-size: 20px;
  font-weight: 500;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: -0.2px;
  color: #888888;
  margin-bottom: 22px;
}

.form {
  display: flex;
  justify-content: space-between;
}

.form-input {
  flex: 3;
  font-size: $medium;
  background-color: #fff;
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
  border-radius: 4px;
  border: solid 1px #e8e8e8;
  font-family: "Montserrat", sans-serif;
}

.form-select {
  flex: 0.75;
  background: $white;
  font-size: $medium * 0.7;
  margin: 0 0.2rem;
  border: 1px solid #e8e8e8;
  border-radius: 4px;
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
  text-align: center;
}

.add-book-btn {
  box-shadow: 0.05rem 0.02rem $black;
  flex: 0.75;
  height: 45px;
  border-radius: 3px;
  font-family: RobotoSlab;
  font-size: $medium * 0.85;
  font-weight: bold;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: 0.5px;
  margin: 0 0.2rem;
  text-align: center;
  color: #ffffff;
  background-color: #0290ff;
  border-color: #0290ff;
}