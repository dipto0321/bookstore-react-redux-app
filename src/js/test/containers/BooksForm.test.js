import React from 'react';
import { shallow } from 'enzyme';
import { BooksForm } from '../../containers/BooksForm';

describe('BooksForm', () => {
  let wrapper;
  let newBook;
  let addBook;
  let inputBook;
  let changeFilter;
  let switchAddBookMode;

  beforeEach(() => {
    addBook = jest.fn();
    switchAddBookMode = jest.fn();
    inputBook = (bookObj) => {
      const {
        title, author, chapters, category,
      } = bookObj;
      wrapper.find('input').at(0).prop('onChange')({
        target: { name: 'title', value: title },
      });

      wrapper.find('input').at(1).prop('onChange')({
        target: { name: 'author', value: author },
      });
      wrapper.find('input').at(2).prop('onChange')({
        target: { name: 'chapters', value: chapters },
      });
      wrapper.find('select').prop('onChange')({
        target: { name: 'category', value: category },
      });
    };

    changeFilter = jest.fn();
    newBook = {
      title: 'Some Title',
      author: 'Some Author',
      chapters: '20',
      category: 'Action',
    };
    wrapper = shallow(
      <BooksForm
        addBook={addBook}
        changeFilter={changeFilter}
        switchAddBookMode={switchAddBookMode}
      />,
    );
  });

  test('it should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should change state during changes in title, author and category inputs', () => {
    const {
      title, author, chapters, category,
    } = newBook;
    inputBook(newBook);
    expect(wrapper.state()).toEqual({
      title, category, author, chapters, error: '',
    });
  });

  test('should show error message if book title is empty or if ', () => {
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {},
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
  });

  test('should call addBook if valid title and category', () => {
    inputBook(newBook);
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {},
    });
    expect(addBook).toHaveBeenLastCalledWith(newBook);
  });
});
