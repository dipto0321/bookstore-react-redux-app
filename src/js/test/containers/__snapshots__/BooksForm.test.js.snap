// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksForm it should render correctly 1`] = `
<div>
  <h2
    className="add-new-title"
  >
    Add new book
  </h2>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      className="form-input"
      onChange={[Function]}
      type="text"
      value=""
    />
    <select
      className="form-select"
      onChange={[Function]}
      value="Select Category"
    >
      <option
        key="Select Category"
        value="Select Category"
      >
        Select Category
      </option>
      <option
        key="Action"
        value="Action"
      >
        Action
      </option>
      <option
        key="Biography"
        value="Biography"
      >
        Biography
      </option>
      <option
        key="History"
        value="History"
      >
        History
      </option>
      <option
        key="Horror"
        value="Horror"
      >
        Horror
      </option>
      <option
        key="Kids"
        value="Kids"
      >
        Kids
      </option>
      <option
        key="Learning"
        value="Learning"
      >
        Learning
      </option>
      <option
        key="Sci-Fi"
        value="Sci-Fi"
      >
        Sci-Fi
      </option>
    </select>
    <button
      className="add-book-btn"
      type="submit"
    >
      Add Book
    </button>
  </form>
</div>
`;
